src/lyrics/lyrics-qt.cc

```diff
@@ -0,0 +1,304 @@
+
+class TextEdit : public QTextEdit
+{
+public:
+    TextEdit (QWidget * parent = nullptr) : QTextEdit (parent) {}
+
+protected:
+    void contextMenuEvent (QContextMenuEvent * event);
+};
+
+class LyricsQt : public GeneralPlugin
+{
+public:
+    static const char * const defaults[];
+    static const PluginPreferences prefs;
+    static const PreferencesWidget widgets[];
+    static constexpr PluginInfo info = {
+        N_("Lyrics"),
+        PACKAGE,
+        nullptr, // about
+        & prefs,
+        PluginQtOnly
+    };
+
+    constexpr LyricsQt () : GeneralPlugin (info, false) {}
+
+    bool init ();
+    void * get_qt_widget ();
+};
+
+EXPORT LyricsQt aud_plugin_instance;
+
+const char * const LyricsQt::defaults[] = {
+    "remote-source", "lyrics.ovh",
+    "enable-file-provider", "TRUE",
+    "enable-cache", "TRUE",
+    "split-title-on-chars", "FALSE",
+    "split-on-chars", "-",
+    "truncate-fields-on-chars", "FALSE",
+    "truncate-on-chars", "|",
+    "use-embedded", "TRUE",
+    nullptr
+};
+
+const PreferencesWidget LyricsQt::widgets[] = {
+    WidgetLabel (N_("<b>General</b>")),
+    WidgetCheck (N_("Split title into artist and title on chars"),
+        WidgetBool (CFG_SECTION, "split-title-on-chars")),
+    WidgetTable ({{split_elements}}, WIDGET_CHILD),
+    WidgetLabel (N_("<b>Sources</b>")),
+    WidgetCheck (N_("Use embedded lyrics (from Lyrics tag)"),
+        WidgetBool (CFG_SECTION, "use-embedded")),
+    WidgetCombo (N_("Fetch lyrics from internet:"),
+        WidgetString (CFG_SECTION, "remote-source"),
+        {{remote_sources}}),
+    WidgetCheck (N_("Store fetched lyrics in local cache"),
+        WidgetBool (CFG_SECTION, "enable-cache")),
+    WidgetLabel (N_("<b>Local Storage</b>")),
+    WidgetCheck (N_("Load lyric files (.lrc) from local storage"),
+        WidgetBool (CFG_SECTION, "enable-file-provider"))
+};
+
+const PluginPreferences LyricsQt::prefs = {{widgets}};
+
+FileProvider file_provider;
+ChartLyricsProvider chart_lyrics_provider;
+LyricsOVHProvider lyrics_ovh_provider;
+
+LyricsState g_state;
+
+static QTextEdit * textedit;
+
+bool LyricsQt::init ()
+{
+    aud_config_set_defaults (CFG_SECTION, defaults);
+    return true;
+}
+
+void update_lyrics_window (const char * title, const char * artist, const char * lyrics)
+{
+    if (! textedit)
+        return;
+
+    textedit->document ()->clear ();
+
+    QTextCursor cursor (textedit->document ());
+    cursor.insertHtml (QString ("<big><b>") + QString (title) + QString ("</b></big>"));
+
+    if (artist)
+        cursor.insertHtml (QString ("<br><i>") + QString (artist) + QString ("</i>"));
+
+    cursor.insertHtml ("<br><br>");
+    cursor.insertText (lyrics);
+}
+
+bool try_parse_json (const Index<char> & buf, const char * key, String & output)
+{
+    QByteArray json = QByteArray (buf.begin (), buf.len ());
+    QJsonDocument doc = QJsonDocument::fromJson (json);
+
+    if (doc.isNull () || ! doc.isObject ())
+        return false;
```

Comment:
Should we call `output = String();` in error cases explicitly?

Reply:
I don't think that's necessary. It's fine to leave `output` unchanged in error cases.